jerry-core/vm/vm.c

```diff
@@ -1186,6 +1239,317 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+            frame_ctx_p->stack_top_p = stack_top_p;
+            frame_ctx_p->call_block_result = block_result;
+            return ECMA_VALUE_UNDEFINED;
+          }
+
+          byte_code_p++;
+          frame_ctx_p->call_operation = VM_NO_EXEC_OP;
+
+          result = *(--stack_top_p);
+          block_result = frame_ctx_p->call_block_result;
+
+          if (ECMA_IS_VALUE_ERROR (result))
+          {
+            goto error;
+          }
+
+          if (!(opcode_data & (VM_OC_PUT_STACK | VM_OC_PUT_BLOCK)))
+          {
+            ecma_fast_free_value (result);
+          }
+          else if (opcode_data & VM_OC_PUT_STACK)
+          {
+            *stack_top_p++ = result;
+          }
+          else
+          {
+            ecma_fast_free_value (block_result);
+            block_result = result;
+          }
+
+          continue;
+        }
+        case VM_OC_CLASS_HERITAGE:
+        {
+          JERRY_ASSERT (byte_code_start_p[0] == CBC_EXT_OPCODE);
+
+          if (byte_code_start_p[1] == CBC_EXT_SUPER_CLASS_CREATE_CONTEXT)
+          {
+            ecma_value_t super_value = *(--stack_top_p);
+            ecma_object_t *super_class_p;
+            branch_offset += (int32_t) (byte_code_start_p - frame_ctx_p->byte_code_start_p);
+
+            JERRY_ASSERT (frame_ctx_p->registers_p + register_end + frame_ctx_p->context_depth == stack_top_p);
+
+            if (ecma_is_value_null (super_value))
+            {
+              super_class_p = ecma_create_object (NULL, 0, ECMA_OBJECT_TYPE_GENERAL);
+            }
+            else
+            {
+              result = ecma_op_to_object (super_value);
+              ecma_free_value (super_value);
+
+              if (ECMA_IS_VALUE_ERROR (result) || !ecma_is_constructor (result))
+              {
+                if (ECMA_IS_VALUE_ERROR (result))
+                {
+                  ecma_free_value (JERRY_CONTEXT (error_value));
+                }
+
+                ecma_free_value (result);
+
+                result = ecma_raise_type_error ("Class extends value is not a constructor or null.");
+                goto error;
+              }
+              else
+              {
+                super_class_p = ecma_get_object_from_value (result);
+              }
+            }
+
+            ecma_object_t *super_env_p = ecma_create_object_lex_env (frame_ctx_p->lex_env_p,
+                                                                     super_class_p,
+                                                                     ECMA_LEXICAL_ENVIRONMENT_SUPER_OBJECT_BOUND);
+
+            ecma_deref_object (super_class_p);
+
+            VM_PLUS_EQUAL_U16 (frame_ctx_p->context_depth, PARSER_SUPER_CLASS_CONTEXT_STACK_ALLOCATION);
+            stack_top_p += PARSER_SUPER_CLASS_CONTEXT_STACK_ALLOCATION;
+
+            stack_top_p[-1] = VM_CREATE_CONTEXT (VM_CONTEXT_SUPER_CLASS, branch_offset);
+
+            frame_ctx_p->lex_env_p = super_env_p;
+          }
+          else
```

Comment:
It seems the two code paths do not share code. Would it be a good idea to introduce two vm opcodes for them?